src/dataflow/src/decode/mod.rs

```diff
@@ -72,42 +78,37 @@ pub fn rewrite_for_upsert(
             Entry::Vacant(vacant) => {
                 // if the key is new, then we know that we always need to ignore the "before" part,
                 // so zero it out
-                vacant.insert(Row::pack_slice(&[after]));
+                vacant.insert(Row::pack_slice(&[after_md]));
 
-                if before.is_null() {
-                    Ok(row)
-                } else {
-                    Ok(Row::pack_slice(&[Datum::Null, after]))
-                }
+                Ok(Row::pack_slice(&[Datum::Null, after_md]))
             }
             Entry::Occupied(mut occupied) => {
-                if occupied.get().iter().next() == Some(before) {
-                    if after.is_null() {
-                        occupied.remove_entry();
-                    } else {
-                        occupied.insert(Row::pack_slice(&[after]));
-                    }
-                    // this matches the modifications we'd make in the next step
-                    Ok(row)
+                let previous_insert = if after.is_null() {
+                    // We are trying to retract something that doesn't exist, so just assume
+                    // that the key is supposed to be empty at this point
+                    let (_k, v) = occupied.remove_entry();
+                    v
                 } else {
-                    let previous_insert = if after.is_null() {
-                        // We are trying to retract something that doesn't exist, so just assume
-                        // that the key is supposed to be empty at this point
-                        let (_k, v) = occupied.remove_entry();
-                        v
-                    } else {
-                        occupied.insert(Row::pack_slice(&[after]))
-                    };
+                    occupied.insert(Row::pack_slice(&[after_md]))
+                };
 
-                    Ok(Row::pack_slice(&[previous_insert.unpack_first(), after]))
-                }
+                Ok(Row::pack_slice(&[previous_insert.unpack_first(), after_md]))
             }
         }
     } else {
         val
     }
 }
 
+fn concat_datum_list<'a, 'b: 'a>(arena: &'a mut Row, left: Datum<'b>, extra: Row) -> Datum<'a> {
+    if matches!(left, Datum::Null) {
+        return left;
+    }
+    let left = left.unwrap_list();
+    arena.push_list(left.into_iter().chain(extra.into_iter()));
```

Comment:
We only have `IntoIterator` implemented for `RowRef`, and I can't find an obvious incantation that is better than `.into_iter()` got get one:

```
error[E0277]: `repr::Row` is not an iterator
   --> src/dataflow/src/decode/mod.rs:108:44
    |
108 |     arena.push_list(left.into_iter().chain(extra));
    |                                            ^^^^^ `repr::Row` is not an iterator
    |
    = help: the trait `Iterator` is not implemented for `repr::Row`
    = note: required because of the requirements on the impl of `IntoIterator` for `repr::Row`
```